tensorflow/contrib/go/graph.go

```diff
@@ -0,0 +1,555 @@
+// across steps.
+func (gr *Graph) Variable(name string, initialData interface{}) (*GraphNode, error) {
+	var dims [][]int64
+
+	ts, err := NewTensor(initialData)
+	if err != nil {
+		return nil, err
+	}
+	gr.variables[name] = ts
+
+	shape := new(pb.TensorShapeProto)
+	if ts.NumDims() == 0 {
+		dims = [][]int64{{1}}
+	} else {
+		dims = ts.Shape()
+	}
+
+	shape.Dim = make([]*pb.TensorShapeProto_Dim, len(dims))
+	for i, dim := range dims {
+		shape.Dim[i] = &pb.TensorShapeProto_Dim{
+			Size: dim[i],
+		}
+	}
+
+	initVal, err := gr.Op("Const", name+"/initial_value", nil, "", map[string]interface{}{
+		"dtype": ts.DataType(),
+		"value": ts,
+		"shape": shape,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	variable, err := gr.Op("Variable", name, nil, "", map[string]interface{}{
+		"dtype":       ts.DataType(),
+		"shape":       shape,
+		"container":   "",
+		"shared_name": "",
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	variable.ref = variable.def
+
+	_, err = gr.Op("Assign", name+"/Assign", []*GraphNode{variable, initVal}, "", map[string]interface{}{
+		"use_locking":    true,
+		"validate_shape": true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	op, err := gr.Op("Identity", name+"/read", []*GraphNode{variable}, "", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	op.ref = variable.def
+
+	return op, nil
+}
+
+// String returns a string representation of this graph, used for debugging
+// proposals.
+func (gr *Graph) String() string {
+	return proto.MarshalTextString(gr.def)
+}
+
+// addInitializationGraphOp add the initialization operation to the graph to
+// cover all the added variables.
+func (gr *Graph) addInitializationGraphOp() {
+	inputs := make([]string, len(gr.variables))
+	i := 0
+	for input := range gr.variables {
+		inputs[i] = "^" + input + "/Assign"
+		i++
+	}
+
+	gr.def.Node = append(gr.def.Node, &pb.NodeDef{
+		Name:  "init",
+		Op:    "NoOp",
+		Input: inputs,
+	})
+}
+
+// Placeholder adds a placeholder to the Graph, a placeholder is an
+// operation that must be fed with data on execution.
+func (gr *Graph) Placeholder(name string, dataType DataType, dims []int64) *GraphNode {
+	op := &GraphNode{
+		outDataTypes: map[string]DataType{
+			name: dataType,
+		},
+		def: &pb.NodeDef{
+			Name: name,
+			Op:   "Placeholder",
+			Attr: make(map[string]*pb.AttrValue),
+		},
+	}
+	op.def.Attr["dtype"] = &pb.AttrValue{
+		Value: &pb.AttrValue_Type{
+			Type: pb.DataType(dataType),
+		},
+	}
+
+	shape := &pb.TensorShapeProto{
+		Dim: make([]*pb.TensorShapeProto_Dim, len(dims)),
+	}
+
+	for i, dim := range dims {
+		shape.Dim[i] = &pb.TensorShapeProto_Dim{
+			Size: dim,
+		}
+	}
+
+	op.def.Attr["shape"] = &pb.AttrValue{
+		Value: &pb.AttrValue_Shape{
+			Shape: shape,
+		},
+	}
+
+	gr.def.Node = append(gr.def.Node, op.def)
+
+	return op
+}
+
+// Str returns the current graph serialized so it can be exported.
+func (gr *Graph) Str() []byte {
+	result, _ := proto.Marshal(gr.def)
```

Comment:
The definition will be always possible to marshal since it is a *pb.GraphDef generated by protobuf. Perhaps it could be any problem on protobuf as it, but this is unlikely.